vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+	    ec.l.simple = {FLT_MAX, 1.f, 0.f};
+	    for (size_t i = 0; i < leaf_labs.size(); i++){
+		    base.predict(ec, b.max_routers + 1 + leaf_labs[i]);
+		    float score = ec.pred.scalar;
+		    if (score > 0)
+			    selected_labs.push_back(leaf_labs[i]);
+	    }
+	    ec.pred.multilabels = preds;
+	    ec.l.multilabels = multilabels;
+
+	    return hamming_loss(ec.l.multilabels.label_v, selected_labs);
+    }
+
+
+    //pick up the "closest" example in the leaf using the score function.
+    int64_t pick_nearest(memory_tree& b, single_learner& base, const uint32_t cn, example& ec)
+    {
+        if (b.nodes[cn].examples_index.size() > 0)
+        {
+            float max_score = -FLT_MAX;
+            int64_t max_pos = -1;
+            for(size_t i = 0; i < b.nodes[cn].examples_index.size(); i++)
+            {
+                float score = 0.f;
+                uint32_t loc = b.nodes[cn].examples_index[i];
+
+                //do not use reward to update memory tree during the very first pass 
+                //(which is for unsupervised training for memory tree)
+                if (b.learn_at_leaf == true && b.current_pass >= 1){ 
+                    float tmp_s = normalized_linear_prod(b, &ec, b.examples[loc]);
+                    example* kprod_ec = &calloc_or_throw<example>();
+                    diag_kronecker_product_test(ec, *b.examples[loc], *kprod_ec, b.oas);
+                    kprod_ec->l.simple = {FLT_MAX, 0., tmp_s};
+                    base.predict(*kprod_ec, b.max_routers);
+                    score = kprod_ec->partial_prediction;
+                    free_example(kprod_ec);
+                }
+                else
+                    score = normalized_linear_prod(b, &ec, b.examples[loc]);
+                
+                if (score > max_score){
+                    max_score = score;
+                    max_pos = (int64_t)loc;
+                }
+            }
+            return max_pos;
+        }
+        else
+            return -1;
+    }
+
+    //for any two examples, use number of overlap labels to indicate the similarity between these two examples. 
+    float get_overlap_from_two_examples(example& ec1, example& ec2){
+        int num_overlap = over_lap(ec1.l.multilabels.label_v, ec2.l.multilabels.label_v);
+        return num_overlap*1.f;
+    }
+
+    //we use F1 score as the reward signal 
+    float F1_score_for_two_examples(example& ec1, example& ec2){
+        float num_overlaps = get_overlap_from_two_examples(ec1, ec2);
+	    float v1 = num_overlaps/(1e-7+ec1.l.multilabels.label_v.size()*1.);
+        float v2 = num_overlaps/(1e-7+ec2.l.multilabels.label_v.size()*1.);
+        if (num_overlaps == 0.f)
+            return 0.f;
+        else
+	    //return v2; //only precision
+            return 2.*(v1*v2/(v1+v2));
+    }
+
+    void predict(memory_tree& b, single_learner& base, example& test_ec)
+    {
+        example& ec = calloc_or_throw<example>();
```

Comment:
Instead of allocating a new example, can we just have a per-reduction example datastructure that is reused?  This is important for efficiency.